tests/test_token_gen.py

```diff
@@ -440,17 +515,31 @@ def test_invalid_arg(self, user_mgt_app, id_token):
             auth.verify_id_token(id_token, app=user_mgt_app)
         assert 'Illegal ID token provided' in str(excinfo.value)
 
-    @pytest.mark.parametrize('id_token', invalid_tokens.values(), ids=list(invalid_tokens))
-    def test_invalid_token(self, user_mgt_app, id_token):
+    @pytest.mark.parametrize('id_token_key', list(invalid_tokens))
+    def test_invalid_token(self, user_mgt_app, id_token_key):
+        id_token = self.invalid_tokens[id_token_key]
+        if _is_emulated() and id_token_key in self.tokens_not_invalid_in_emulator:
+            # These tokens won't be invalid when using the emulator, check them like valid tokens.
```

Comment:
There's quite a bit of duplication and commentary here. Let's try to reduce that as much as possible. For example:

```
if _is_emulated() and id_token_key in self.tokens_not_invalid_in_emulator:
    return _assert_valid_in_emulator(id_token, user_mgt_app)
```

Then you can use the same helper function in places like `test_expired_token`.